경북대 BE_유우석 : Step3 기능 구현

milk-stone opened the pull request:
안녕하십니까 신종화 멘토님, 3단계 기능을 구현하여 리뷰 요청드립니다.

구현하고자 하는 기능은 `README.md`에 정리해두었고, JDBC 기술 중 JdbcClient 를 사용하여 구현하였습니다.

이에 앞서, 지난 Step2 리뷰 관련해서 리뷰 요청 기한을 지키지 못한 점에 대해 진심으로 사과드립니다.

특히 멘토님의 배려로 마감 기한을 하루 연장해 주셨음에도 불구하고, 그 약속마저 지키지 못해 정말 죄송하고 부끄러운 마음입니다. 저의 불찰과 안일한 시간 관리로 인해 멘토님께 큰 실망감을 드렸던 것 같습니다. 이번 일을 계기로 카테캠에 임하는 자세를 다시 한 번 점검하여 앞으로는 이런 일이 절대 일어나지 않도록 각별히 주의하겠습니다.

milk-stone committed:
docs(readme): Step3에서 구현할 기능 정리

milk-stone committed:
chore(application): H2 데이터베이스 연동 설정

- application.properties : h2 database 활성화, url, 드라이버 클래스, 유저 네임 등 설정
- schema.sql : 스키마 제작 스크립트
- data.sql : 데이터 스크립트, 실행 시 스키마에 맞는 데이터 10개를 생성

milk-stone committed:
feat(repository): 상품 단건 조회, 모든 상품 조회, 상품 등록, 상품 삭제, 상품 수정 로직 모두 JdbcClient 를 사용하도록 변경

- update 함수의 반환형 int로 변경, affectedRows를 반환함으로써 Service 단에서 Update의 성공 여부를 판단할 수 있게 함.
- 다른 API는 메서드 입력 출력 구조 변경 없이 내부 동작 코드만 JdbcClient를 이용하는 코드로 수정하였음.
- 항상 Application 실행 시, 더미 데이터 10개를 생성하므로 nextId = 11L; 로 설정하여 id 값이 11부터 증가하도록 설계 하였음.

milk-stone committed:
feat: Update 메서드가 ProductResponse가 아닌 int를 반환함에 따라 update 메서드와 관련된 service, controller 메서드 수정

milk-stone committed:
Feat(products.html): 서버 사이드 렌더링 페이지 수정

- 상품 단건 조회 기능을 누락했던 것을 추가하였음.
    - 작업 란의 파란색 눈 모양을 클릭하면 해당 상품의 상세 정보를 열람할 수 있음.
- 이미지 링크를 보여주던 Column 이름이 작업으로 보여졌던 현상 수정

jjongwa commented on src/main/resources/data.sql:
schema.sql로 스키마 생성 & data.sql로 기본 데이터 세팅까지 구성해 주셨군요~! 💯

jjongwa commented on src/main/resources/application.properties:
```diff
@@ -1 +1,13 @@
+spring.datasource.password=
+#spring.sql.init.schema-locations=schema.sql
+#spring.sql.init.data-locations=data.sql
+spring.sql.init.mode=always
```
init.mode=always로 항상 초기화도 좋습니다~ 👏

milk-stone replied:
감사합니다!

jjongwa commented on src/main/resources/application.properties:
```diff
@@ -1 +1,13 @@
 spring.application.name=spring-gift
+# h2-console ??? ??
+spring.h2.console.enabled=true
+# ??? ??(DataSource) ??
+spring.datasource.url=jdbc:h2:mem:gift
+spring.datasource.driverClassName=org.h2.Driver
+# H2 ?? ??? ??? 'sa'
+spring.datasource.username=sa
+# H2 ?? ????? ??
+spring.datasource.password=
+#spring.sql.init.schema-locations=schema.sql
+#spring.sql.init.data-locations=data.sql
```
불필요한 주석은 제거해 주세요~!
spring.sql.init.schema-locations=schema.sql 같은 설정을 나중에 쓰겠다 하는 의미더라도 여기에 굳이 남겨둘 필요는 없습니다! 그래서 커밋이 있는 거니까요.!

(H2 ?? ????? ?? 요런건 뭔가요?)

milk-stone replied:
- "# h2-console 활성화"
- "# 데이터 소스(DataSource) 설정"
- "# H2 기본 사용자 이름 'sa'"
- "# H2 기본 비밀번호 공란"

와 같은 주석들이었는데, 한글 인코딩이 깨졌던 것 같습니다 ㅠㅠ

application.properties의 각 행들은 spring.datasource.username 등 코드만 봐도 설명이 되므로, 주석은 모두 지운 후 커밋하겠습니다!
또한, 주석이 필요한 상황에서는 한글 인코딩이 혹시나 깨지지는 않았는지 한 번 더 확인하는 습관을 들이도록 하겠습니다!

jjongwa commented on src/main/java/gift/product/repository/ProductRepository.java:
```diff
@@ -3,47 +3,69 @@
 @Repository
 public class ProductRepository {
-    private final Map<Long, Product> products = new HashMap<>();
+    private final JdbcClient jdbcClient;
```
jdbcClient를 사용해 주셨네요!

다른 대안으로는 JdbcTemplate, SimpleJdbcinsert 등이 있을 텐데요,
그 중 jdbcClient를 선택한 이유는 무엇인가요?

milk-stone replied:
이전에 1단계 과제를 제출할 때, JdbcTemplate를 사용했었습니다. 이번 과제에서는 지난번에 사용해본 JdbcTemplate를 제외한 두 기술 중에서 고르려고 했습니다. 그 중 jdbcClient를 사용하는 코드가 저에게 더 직관적인 코드로 느껴져서 선택했습니다!

jjongwa replied:
오호.. 좋습니다~!

jjongwa commented on src/main/java/gift/product/repository/ProductRepository.java:
```diff
@@ -3,47 +3,69 @@
 
     public Optional<Product> save(ProductRequest request){
-        if (products.containsKey(nextId)){
-            return Optional.empty();
-        }
         Product product = new Product(nextId, request.name(), request.price(), request.imageUrl());
-        products.put(nextId, product);
+
+        String insertSql = "INSERT INTO product (id, name, price, image_url) VALUES (:id, :name, :price, :imageUrl)";
+        jdbcClient.sql(insertSql)
+                .param("id", nextId)
+                .param("name", request.name())
+                .param("price", request.price())
+                .param("imageUrl", request.imageUrl())
+                .update();
         nextId = nextId + 1;
```
저장방식을 DB로 변경했지만, 아직 ID 생성은 Repository 단에서 담당하고 있네요.
이러한 경우 여전히 동시성 문제가 발생할 수 있습니다.

ID 생성의 책임까지 DB에 맡기도록 변경해 볼까요?

milk-stone replied:
넵 알겠습니다! 코드 수정 해보겠습니다!

jjongwa commented on src/main/java/gift/product/repository/ProductRepository.java:
```diff
@@ -3,47 +3,69 @@
-        }
         Product product = new Product(nextId, request.name(), request.price(), request.imageUrl());
-        products.put(nextId, product);
+
```
save 메서드가 ProductRequest DTO를 직접 받고 있네요.
Request 라는 네이밍에서 알 수 있듯이 해당 DTO는 표현(Controller) 계층에서부터 영속성(Repository) 계층까지 모두 넘나들고 있는데요, 

이렇게 어짜피 모든 계층을 자유롭게 지나다니는 객체가 있다면 Controller- Service - Repository 계층 구조를 나눈 이유가 없어지지 않을까요?

milk-stone replied:
앗 맞네요 감사합니다! DTO의 존재 이유를 잘못 생각하고 있었던 것 같습니다.

- 기존 생각
  - Product와 같은 엔티티 객체를 직접 파라미터로 넣어 사용하면, 다른 계층에서 엔티티 내용을 변질시킬 수 있습니다. 그렇기에, 변경되지 않는 내용만 담는 DTO를 만들어서 계층간에 정보를 전달한다고 생각했습니다. DTO가 특정 계층 사이의 전달 객체라고 생각하지 않고, 단순히 내용만 담아놓은 불변 객체라고 생각했었습니다.

- 멘토님의 말을 들은 직후의 생각
  - 계층 구조를 나는 것이 한 코드의 영향력을 제한하기 역할도 수행하는데, DTO를 전 계층에 걸쳐서 사용하게 되면, 해당 DTO를 수정할 때, 전 계층의 코드를 전부 확인하여 고쳐야할 것 같다는 생각을 했습니다.
  - 따라서, DTO의 역할을 외부에서 들어오는 Json 데이터를 파싱하여 저장하고, Controller 단에서 Service 단으로 내용을 전송하는 역할로만 제한해야겠다고 생각했습니다.

- 코드 수정 방향
  - DTO가 Repository 단까지 사용되는 코드 부분을 수정하겠습니다.
  - Service와 Repository는 모두 Product 엔티티를 import 하므로, Service ←→ Repository 단 사이의 데이터 전송은 Product 엔티티 자체를 이용하도록 수정해보겠습니다.
  
좋은 피드백 감사합니다! 덕분에 다양한 생각을 해볼 수 있었던 것 같습니다! :)

jjongwa replied:
👏 👏 아주 좋습니다~!
> Product와 같은 엔티티 객체를 직접 파라미터로 넣어 사용하면, 다른 계층에서 엔티티 내용을 변질시킬 수 있습니다. 

요런 문제는 Product가 불변 객체여도 해결 가능하겠네요! (내부 값이 변하지 않는다 -> setter를 사용하지 않는다)
왜 setter 사용을 지양하라고 하는지 어느 정도 느낌이 오실 수도 있겠는걸요?

jjongwa commented on src/main/java/gift/product/repository/ProductRepository.java:
```diff
@@ -3,47 +3,69 @@
 
     public Optional<List<Product>> getAll(){
-        List<Product> productList = new ArrayList<>(products.values());
-        return Optional.of(productList);
+        String sql = "SELECT * FROM product";
+        return Optional.of(
+                jdbcClient.sql(sql)
+                .query(Product.class)
+                .list()
+        );
```
리스트 반환의 경우 null 이나 Optional.empty() 보단 비어 있는 컬렉션(Collections.emptyList()) 을 반환하는 것이 조금 더 표준적인 패턴입니다!

(만약 Optional.of() 안에 빈 리스트가 들어가 있다면 empty()는 무엇을 반환할까요?)

milk-stone replied:
저도 List<Product> 객체를 Optional<> 로 감싸는게 맞는 방법일까 고민했었는데 표준적인 패턴을 알려주셔서 감사합니다! 수정해서 제출하도록 하겠습니다!


> 만약 Optional.of() 안에 빈 리스트가 들어가 있다면 empty()는 무엇을 반환할까요?
- 빈 리스트일 경우 Null로 판단하여 예외 처리를 해야하지만, isEmpty() 에서는 비지 않았다는 결과가 나올 것 같습니다.

jjongwa reviewed (changes requested):
3단계 빠르게 잘 구현해 주셨습니다!

간단한 몇가지 리뷰 남겼으니 반영 후 다시 요청 주세요~!

깔끔한 코드를 제출하는 것도 중요하지만, 가장 중요한건 주어진 시간 내에 일단 돌아가는 코드를 완성하는 겁니다!
앞으로의 미션에서는 제출 기한 -> 현업에서의 due date 라고 생각하고 임해 주세요!!!
(프로젝트 오픈 예정일이 지나도 개발이 안올라온다? 그럼 다음에 아무도 일 안맡깁니다..)

milk-stone committed:
fix(application.properties): 알아볼 수 없는 문자 삭제

인코딩 방식이 잘못됐는지, 입력 할 당시에는 한글로 보였는데 어느새 글자가 깨져있었습니다.
커밋 한 이후에 커밋 내역을 살펴보면서 의도대로 커밋이 됐는지 확인하는 습관을 들여야 겠다고 생각했습니다.

milk-stone committed:
refactor: 2가지 변경 사항

### ID 생성을 DB에 위임하도록 수정
`schema.sql`
- id 컬럼에 AUTO_INCREMENT 특성을 추가하여 DB가 ID를 관리하게 함.

`data.sql`
- id 값 관리는 DB가 하므로, 초기 데이터 생성 INSERT 문의 ID 컬럼 삭제

`Product.java`
- All-args 생성자 -> protected No-args 생성자 + id를 제외한 필드를 입력받는 생성자 로 수정
  - 수정 이유 <br>
    더 이상 id 필드를 직접 설정하지 않기에 생성자에서 id 필드를 제외해야 했고, (name, price, imageUrl) 만 입력받는 생성자만 두려고 하였음.
    하지만, 인터넷에 검색해보니 protected 접근 단계의 인자 없는 기본 생성자를 같이 선언하는 것이 좋다는 글을 읽어 위와 같이 구성하였음.
    - protect No-args 생성자 : 프레임 워크 호환용
    - public (name, price, imageUrl) 생성자 : id를 제외한 생성자
- 이렇게 생성자를 구성할 시, Setter가 없는 상황인 지금 Product의 id 필드에 값을 부여할 방법이 존재하지 않음.
  - 따라서, Product 객체 완성을 위한 setId() 세터를 따로 두었음.

### ProductRequest 가 Repository에 접근하지 않도록 수정

`ProductService.java`
- Service 단에서 DTO의 정보를 이용하여 Product 객체 생성 후 Repository에 Product 엔티티를 전달하도록 수정.

`ProductRepository.java`
- nextID 필드 삭제 : DB가 ID를 관리하기에 더 이상 필요하지 않음.
- insert 문의 id 컬럼 삭제.

milk-stone committed:
"refactor: 1가지 변경 사항 적용

### ProductUpdateRequest DTO가 Repository 까지 전달되는 현상 수정

- `Product.java`
  - name, price, imageurl을 인자로 받는 update 메서드 작성
    - 전에는 DTO를 직접 인자로 넣어 작성하였는데, DTO를 직접 넣는 방식은 좋지 않다고 생각하여 이 메서드를 삭제하였었음.
    - 그러나, DTO를 import 하지 않고, (name, price, imageUrl)를 인자로 받는 update 메서드를 작성하면, DTO를 import 하지 않는 선에서 Product의 내용을 업데이트 할 수 있음.

- `ProductRepository.java`
  - Product를 인자로 받아 수정하는 Sql문을 작성하도록 수정.

- `ProductService.java`
  - 조회 - 수정 - DB저장 순서대로 작동하도록 코드 수정."

milk-stone committed:
refactor: 1가지 변경 사항 적용

### Optional<컬랙션> 을 반환하던 것을 비어 있는 컬랙션을 반환하도록 수정

- `ProductRepository.java`
  - Optional으로 감싸던 것을 삭제.

- `ProductService.java`
  - JdbcClient가 null을 반환하였을 경우 (진짜 조회에 실패한 경우) : 예외 처리
  - JdbcClient가 빈 배열을 반환하였을 경우 (조회는 성공하였으나, 조회된 것이 없는 경우) : null 반환
  - 정상 작동한 경우 : 조회된 Product를 ProductResponse로 매핑하여 반환

jjongwa commented on src/main/java/gift/product/repository/ProductRepository.java:
```diff
@@ -1,49 +1,59 @@
-        Product product = new Product(nextId, request.name(), request.price(), request.imageUrl());
-        products.put(nextId, product);
-        nextId = nextId + 1;
+    public Optional<Product> save(Product product){
+        String insertSql = "INSERT INTO product (name, price, image_url) VALUES (:name, :price, :imageUrl)";
+        jdbcClient.sql(insertSql)
+                .param("name", product.getName())
+                .param("price", product.getPrice())
+                .param("imageUrl", product.getImageUrl())
+                .update();
         return Optional.of(product);
```
직관적으로 느껴진 부분은 요 부분이겠죠?ㅎㅎ

jjongwa commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,10 @@
+
+-- products 테이블 생성
+CREATE TABLE product (
+    id BIGINT AUTO_INCREMENT PRIMARY KEY,
```
👍 👍

jjongwa commented on src/main/java/gift/product/repository/ProductRepository.java:
```diff
@@ -1,49 +1,59 @@
-    public Optional<List<Product>> getAll(){
-        List<Product> productList = new ArrayList<>(products.values());
-        return Optional.of(productList);
+    public List<Product> getAll(){
+        String sql = "SELECT * FROM product";
+        return jdbcClient.sql(sql)
+                .query(Product.class)
+                .list();
```
👍

jjongwa reviewed (approved):
고생 많으셨어요~! 이번 미션은 여기서 마무리 하곘습니다!
단계를 거쳐 가면서 여러 방면에서 성장하는 모습이 보였던 것 같네요~!

지금 진행하고 있는 카테캠 기간이 어쩌면 졸업 전 폭발적으로 성장할 수 있는 마지막 기회일 수 있어요.!
뒤돌아 봤을 때 후회 하지 않을 수 있도록 다음 미션에서도 이대로 쭉쭉 달려가시길 바라겠습니다~!! 🏃 🏃
